File: python-backend/minimax.py
from evaluate_func_lists import evaluate_function
from constants import white, black, width, height
from possible_moves import get_possible_moves


def minimax(my_dots, another_dots, width, height, maximizing, depth):
    if depth == 0:
        return evaluate_function(my_dots, another_dots, width, height)

    possible_moves = get_possible_moves(my_dots, another_dots, width, height)

    if maximizing:
        best_score = float("-inf")
        for move, changes in possible_moves.items():
            add_changes(move, changes, my_dots, another_dots)
            print("here")

            score = minimax(
                another_dots,
                my_dots,
                width,
                height,
                maximizing=False,
                depth=depth - 1,
            )
            best_score = max(score, best_score)

            undo_changes(move, changes, my_dots, another_dots)

    else:
        best_score = float("inf")
        for move, changes in possible_moves.items():
            add_changes(move, changes, my_dots, another_dots)

            score = minimax(
                another_dots,
                my_dots,
                width,
                height,
                maximizing=True,
                depth=depth - 1,
            )
            best_score = min(score, best_score)

            undo_changes(move, changes, my_dots, another_dots)

    return best_score


def add_changes(move, changes, my_dots, another_dots):
    for change in changes:
        my_dots.append(change)
        another_dots.remove(change)
    my_dots.append(move)


def undo_changes(move, changes, my_dots, another_dots):
    for change in changes:
        my_dots.remove(change)
        another_dots.append(change)
    my_dots.remove(move)


def get_bot_move(bot_dots, another_dots, width, height):
    possible_moves = get_possible_moves(bot_dots, another_dots, width, height)
    best_move = None
    best_move_changes = None
    best_score = float("-inf")
    for move, changes in possible_moves.items():
        add_changes(move, changes, bot_dots, another_dots)

        score = minimax(
            another_dots, bot_dots, width, height, maximizing=False, depth=1
        )
        if best_score < score:
            best_score = score
            best_move = move
            best_move_changes = changes

        undo_changes(move, changes, bot_dots, another_dots)

    return best_move, best_move_changes, possible_moves.keys()

File: tau-prolog-backend/minimax.pl
:- consult('possible_moves.pl').
:- consult('evaluate_functions.pl').

/* minimax(+MyDots, +AnotherDots, +Width, +Height, +Maximizing, +Depth, -BestScore) */
minimax(MyDots, AnotherDots, Width, Height, _, Maximizing, 0, BestScore) :- 
    evaluate_function(MyDots, AnotherDots, Width, Height, Maximizing, BestScore), !.

minimax(MyDots, AnotherDots, Width, Height, Holes, maximizing, Depth, BestScore) :- 
    NewDepth is Depth - 1,
    get_possible_moves(AnotherDots, MyDots, Width, Height, Holes, Moves),
    findall(Score, (member(Move, Moves), add_changes(Move, MyDots, AnotherDots, NewMyDots, NewAnotherDots), 
                   minimax(NewAnotherDots, NewMyDots, Width, Height, Holes, minimizing, NewDepth, Score)), Scores),
    (   Scores = []
    ->  get_possible_moves(MyDots, AnotherDots, Width, Height, Holes, Moves2),
        findall(Score, (member(Move, Moves2), add_changes(Move, MyDots, AnotherDots, NewMyDots, NewAnotherDots), 
                   minimax(NewMyDots, NewAnotherDots, Width, Height, Holes, maximizing, NewDepth, Score)), Scores2),
        (   Scores2 = []
        ->  
            who_won(MyDots, AnotherDots, maximizing, BestScore)
        ;   max_list(Scores, BestScore)
        )
    ;   max_list(Scores, BestScore1),
        BestScore is BestScore1 + 100 % add bonus because opponent had no moves
    ) 
    , !.

minimax(MyDots, AnotherDots, Width, Height, Holes, minimizing, Depth, BestScore) :- 
    NewDepth is Depth - 1,
    get_possible_moves(AnotherDots, MyDots, Width, Height, Holes, Moves),
    findall(Score, (member(Move, Moves), add_changes(Move, MyDots, AnotherDots, NewMyDots, NewAnotherDots), 
                   minimax(NewAnotherDots, NewMyDots, Width, Height, Holes, maximizing, NewDepth, Score)), Scores),
    (   Scores = []
    ->  get_possible_moves(MyDots, AnotherDots, Width, Height, Holes, Moves2),
        findall(Score, (member(Move, Moves2), add_changes(Move, MyDots, AnotherDots, NewMyDots, NewAnotherDots), 
            minimax(NewMyDots, NewAnotherDots, Width, Height, Holes, minimizing, NewDepth, Score)), Scores2),
        (   Scores2 = []
        ->  who_won(MyDots, AnotherDots, minimizing, BestScore)
        ;   min_list(Scores, BestScore)
        )
    ;   min_list(Scores, BestScore1),
        BestScore is BestScore1 - 100 % add bonus because opponent had no moves
    )
    , !.

/* add_changes(+Move, +MyDots, +AnotherDots, -NewMyDots, -NewAnotherDots) */
add_changes(move(Move, Changes), MyDots, AnotherDots, NewMyDots, NewAnotherDots) :-
    append(Changes, [Move | MyDots], NewMyDots),
    substract_list(AnotherDots, Changes, NewAnotherDots), !.
    

/* substract_list(+List1, +List2, -Result) */
substract_list([], _, []) :- !.
substract_list([Head | List1], List2, Result) :- 
    member(Head, List2),
    substract_list(List1, List2, Result), !.

substract_list([Head1 | List1], List2, [Head1 | Result]) :- 
    substract_list(List1, List2, Result), !.